lib/NodeUtils.js

```diff
@@ -116,6 +116,82 @@ function attrname(a) {
   return a.name;
 }
 
+/**
+ * Escapes matching closing tag in a raw text.
+ *
+ * For example, given `<style>#text(</style><script></script>)</style>`,
+ * the parent tag would by "style" and the raw text is
+ * "</style><script></script>". If we come across a matching closing tag
+ * (in out case `</style>`) - replace `<` with `&lt;` to avoid unexpected
+ * and unsafe behavior after de-serialization.
+ */
+function escapeMatchingClosingTag(rawText, parentTag) {
+  const regexp = new RegExp('</' + parentTag, 'ig');
+
+  if (!regexp.test(rawText)) {
+    return rawText; // fast path
+  }
+
+  regexp.lastIndex = 0; // reset an index to start from the beginning of a string
+  const matches = rawText.matchAll(regexp);
+
+  let result = '';
+  let lastIndex = 0;
+  for (const match of matches) {
+    result += rawText.substring(lastIndex, match.index) + '&lt;';
+    // Skip over the `<` char, since we've replaced it above.
+    // Start the next iteration with a char following the `<` one.
+    lastIndex = match.index + 1;
+  }
+  // Add remaining contents of a string after the last `<` char.
+  if (lastIndex < rawText.length) {
+    result += rawText.substring(lastIndex);
+  }
+  return result;
+}
+
+const CLOSING_COMMENT_REGEXP = /--!?>/;
+const CLOSING_COMMENT_REGEXP_GLOBAL = /--!?>/g;
+
+/**
+ * Escapes closing comment tag in a comment content.
+ *
+ * For example, given `#comment('-->')`, the content of a comment would be
+ * updated to `--&gt;` to avoid unexpected and unsafe behavior after
+ * de-serialization.
+ */
+function escapeClosingCommentTag(rawContent) {
```

Comment:
There's an off-by-one error in this function. 

```
escapeClosingCommentTag("asd-->a<script>")
// returns 'asd-->&gt;<script>'
```

The comment end still persists and erroneously the next character (letter `"a"`) is being replaced with `&gt;` instead.

I also wonder whether this function can be replaced with just a `rawContent.replaceAll` but I might be missing some edge cases here.

Reply:
@securityMB great catch, thanks for looking into this! 👍

I've applied the changes in https://github.com/angular/domino/pull/18/commits/9bf2b98233b61d1c92eca4fb4b78afb3e9786572 and also extended tests to verify HTML output (some tests previously only ran a check in a browser).

I think using `replaceAll` is a good idea, please let me know if you see any edge-cases.